sycl/source/handler.cpp

```diff
@@ -842,6 +857,280 @@ void handler::ext_oneapi_memset2d_impl(void *Dest, size_t DestPitch, int Value,
+  MImpl->MImageFormat = PiFormat;
+  MImpl->MImageCopyFlags = PiCopyFlags;
+  setType(detail::CG::CopyImage);
+}
+
+void handler::ext_oneapi_copy(
+    void *Src, sycl::range<3> SrcOffset, void *Dest, sycl::range<3> DestOffset,
+    const ext::oneapi::experimental::image_descriptor &DeviceImgDesc,
+    size_t DeviceRowPitch, sycl::range<3> HostExtent,
+    sycl::range<3> CopyExtent) {
+  MSrcPtr = Src;
+  MDstPtr = Dest;
+
+  sycl::detail::pi::PiMemImageDesc PiDesc = {};
+  PiDesc.image_width = DeviceImgDesc.width;
+  PiDesc.image_height = DeviceImgDesc.height;
+  PiDesc.image_depth = DeviceImgDesc.depth;
+  PiDesc.image_type = DeviceImgDesc.depth > 0
+                          ? PI_MEM_TYPE_IMAGE3D
+                          : (DeviceImgDesc.height > 0 ? PI_MEM_TYPE_IMAGE2D
+                                                      : PI_MEM_TYPE_IMAGE1D);
+
+  sycl::detail::pi::PiMemImageFormat PiFormat;
+  PiFormat.image_channel_data_type =
+      sycl::_V1::detail::convertChannelType(DeviceImgDesc.channel_type);
+  PiFormat.image_channel_order =
+      sycl::_V1::detail::convertChannelOrder(DeviceImgDesc.channel_order);
+
+  // Flags
+  sycl::detail::pi::PiImageCopyFlags PiCopyFlags;
+  sycl::usm::alloc DstPtrType = get_pointer_type(Dest, MQueue->get_context());
+  sycl::usm::alloc SrcPtrType = get_pointer_type(Src, MQueue->get_context());
+  if (DstPtrType == sycl::usm::alloc::device) {
+    // Dest is on device
+    if (SrcPtrType == sycl::usm::alloc::device) {
+      PiCopyFlags =
+          sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_DEVICE_TO_DEVICE;
+    } else if (SrcPtrType == sycl::usm::alloc::host ||
+               SrcPtrType == sycl::usm::alloc::unknown) {
+      PiCopyFlags =
+          sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_HOST_TO_DEVICE;
+    } else {
+      assert(false && "Unknown copy source location");
+    }
+  } else if (DstPtrType == sycl::usm::alloc::host ||
+             DstPtrType == sycl::usm::alloc::unknown) {
+    // Dest is on host
+    if (SrcPtrType == sycl::usm::alloc::device) {
+      PiCopyFlags =
+          sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_DEVICE_TO_HOST;
+    } else if (SrcPtrType == sycl::usm::alloc::host ||
+               SrcPtrType == sycl::usm::alloc::unknown) {
+      assert(false && "Cannot copy image from host to host");
+    } else {
+      assert(false && "Unknown copy source location");
+    }
+  } else {
+    assert(false && "Unknown copy destination location");
+  }
+
+  MImpl->MSrcOffset = {SrcOffset[0], SrcOffset[1], SrcOffset[2]};
+  MImpl->MDestOffset = {DestOffset[0], DestOffset[1], DestOffset[2]};
+  MImpl->MHostExtent = {HostExtent[0], HostExtent[1], HostExtent[2]};
+  MImpl->MCopyExtent = {CopyExtent[0], CopyExtent[1], CopyExtent[2]};
+  MImpl->MImageDesc = PiDesc;
+  MImpl->MImageDesc.image_row_pitch = DeviceRowPitch;
+  MImpl->MImageFormat = PiFormat;
+  MImpl->MImageCopyFlags = PiCopyFlags;
+  setType(detail::CG::CopyImage);
```

Comment:
It looks like this has a lot of code in common with the previous function. Could we move the common code to a common function or maybe have one of these call the other?

Reply:
I've moved getting the copy flags into a separate helper functions. Setting the other values could maybe be moved into separate functions as well, but we often use different descriptors and assignments so it might get convoluted if we do so.